Merge dashboard-ansible code into ceph-ansible

b-ranto opened the pull request:
This PR merges the dashboard instllation code into ceph-ansible. The code only supports containerized deployment and the PR creates four new roles: ceph-grafana, ceph-dashboard, ceph-prometheus and ceph-node-exporter.

Edit (@zmc):
To-do:
- [x] `purge-cluster.yml`
- [x] `site-container.yml.sample`
- [x] `purge-docker-cluster.yml`
- [x] Use packaged dashboards
- [x] Squash and/or clean up commits
- [x] Add podman support

leseb commented on roles/ceph-dashboard/tasks/configure_dashboard.yml:
```diff
@@ -0,0 +1,140 @@
+    - not dashboard.key or not dashboard.crt
+
+- name: Import dashboard certificate file
+  command: "{{ mgr_prefix }} ceph config-key set mgr/dashboard/crt -i /etc/ceph/ceph-dashboard.crt"
```
Use `changed_when: false` for `command` calls.

leseb commented on roles/ceph-dashboard/tasks/main.yml:
```diff
@@ -0,0 +1,6 @@
+---
+- include: configure_dashboard.yml
```
use `include_tasks` instead.

leseb commented on roles/ceph-grafana/tasks/configure_firewall.yml:
```diff
@@ -0,0 +1,17 @@
+---
+- name: Check firewalld status
+  shell: "systemctl show firewalld | grep UnitFileState"
```
see `ceph-infra` role for fw rules instead

leseb commented on roles/ceph-grafana/tasks/main.yml:
```diff
@@ -0,0 +1,7 @@
+---
+- include: setup_container.yml
```
use `include_tasks`

leseb replied:
Also, name your include.

```
- name: include setup_container.yml
  include_tasks: setup_container.yml
```

leseb commented on roles/ceph-node-exporter/handlers/main.yml:
```diff
@@ -0,0 +1,9 @@
+---
```
see the role `ceph-handler` for handlers.

leseb commented on roles/ceph-node-exporter/tasks/main.yml:
```diff
@@ -0,0 +1,22 @@
+---
+- include: setup_container.yml
```
name your include and use `include_tasks`

leseb commented on roles/ceph-node-exporter/tasks/main.yml:
```diff
@@ -0,0 +1,22 @@
+---
+- include: setup_container.yml
+
+- name: Check firewalld status
```
ceph-infra role for fw

leseb commented on site.yml.sample:
```diff
@@ -463,3 +467,28 @@
+        name: ceph-grafana
+      when: dashboard_enabled
+
+- hosts: mgrs[0]
```
There is no mgrs[0] by default anymore, now we collocate mon/mgr by default. So 3 mons = 3 mgrs

b-ranto commented on roles/ceph-container-common/tasks/main.yml:
```diff
@@ -18,10 +18,17 @@
   include_tasks: fetch_image.yml
   tags:
     - fetch_container_image
+  when: containerized_deployment
```
This is trickier. We only support containerized deployment in dashboard-ansible. I need to include this role to get docker installed even if ceph itself is not containerized. This allows me to do that.

zmc commented:
While @b-ranto is out, I'm taking over this work. I've addressed most, if not all, of the feedback items, as well as fixed several issues I noticed while testing it out. I do currently have a successful deployment, but there are some outstanding issues. I'll edit the initial comment on the PR with a task list, and push my updates.

b-ranto commented on roles/ceph-grafana/files/dashboards/ceph-cluster.json:
```diff
@@ -51,7 +51,7 @@
         "rgba(237, 129, 40, 0.9)",
         "rgb(255, 0, 0)"
       ],
-      "datasource": "$datasource",
```
FYI: This was intentional, with `$datasource` here, the datasource was actually selectable in grafana. We should probably drop the dashboards from the PR altogether and use those that are provided in ceph's `ceph-grafana-dashboards` package. Having them as part of the deployment was sort of a hack for luminous where these dashboards did not exist upstream, yet. Alternatively, we can have them as back-up in case the package does not exist (that was my original plan, the code should be ready for this, it just does not install the package, yet).

zmc replied:
Oh, okay - it's just that none of the dashboards were working by default. I'll update the PR to use the packaged dashboards. It does look like they use a templated datasource name as well, so I'll see about fixing the default behavior.

I'd tentatively suggest that we shouldn't store them in two places, for maintenance reasons. It shouldn't be hard to backport the dashboards to luminous if we need to.

b-ranto replied:
It was a downstream hack so we probably don't need to care about it upstream and we can just drop the dashboards and install the package instead. The official dashboard v2 support is dedicated for nautilus+.

It seems weird that they did not work, what error did you get? What grafana version were you using? Did you by any chance hit this bug? https://github.com/grafana/grafana/issues/14239

zmc replied:
It was/is v5.4.2 - I actually can't remember if it produced an error - the main symptom was blank graphs. Once I get the purge playbook working I'll tear down my current setup and dig a little deeper.

zmc replied:
Update: I switched to using the packaged dashboards. It looks like the failure in datasource name templating only happens when visiting grafana directly without being signed in; once I logged in as admin, they worked properly. When viewing from the dashboard, they worked as well (even before signing in on the other grafana tab).

zmc commented:
Looking at `site-container.yml.sample` and `purge-docker-cluster.yml`, I can't help but wish the various playbooks could share some code, and have access to var defaults. I wonder if it would ever make sense to move purge logic into relevant roles, and have their operation controlled by vars.

zmc commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Dashboard: Ceph-Ansible deployment](https://trello.com/c/UUVmek0P/89-dashboard-ceph-ansible-deployment)

b-ranto commented on roles/ceph-infra/tasks/configure_firewall.yml:
```diff
@@ -175,6 +175,15 @@
+      source: "{{ public_network }}"
+      permanent: true
+      immediate: true
+      state: enabled
```
We should probably add `when: dashboard_enabled`, here.

b-ranto commented on roles/ceph-infra/tasks/configure_firewall.yml:
```diff
@@ -179,7 +179,14 @@
+  - name: Open ports for node_exporter
+    firewalld:
+      port: "9100/tcp"
+      zone: "{{ ceph_dashboard_firewall_zone }}"
```
The same, here for node_exporter (`when: dashboard_enabled`) and any other dashboard-related ports.

b-ranto commented on roles/ceph-grafana/tasks/configure_grafana.yml:
```diff
@@ -72,13 +72,6 @@
-- name: Store the dashboard files
-  copy:
-    src: "files/dashboards/"
-    dest: "{{ grafana.dashboards_path }}"
```
We can probably also remove the `grafana.dashboard_path` variable in this commit.

zmc replied:
Yeah, this is all gone now

b-ranto commented on group_vars/all.yml.sample:
```diff
@@ -662,6 +663,17 @@ dummy:
+#############
+# DASHBOARD #
+#############
+#dashboard_enabled: True
```
This reminds me that we should probably make this default to `False` since dashboard won't be (fully) usable without a `ceph-grafana` node.

zmc commented:
Okay, I think this is ready to be merged. To fully tie everything together, it will require some changes to ceph's dashboard module, which is in-progress. When there's a PR for that, I'll link it here for posterity. It should be in the next day or so.

What it's going to do, is cause the dashboard to push the dashboard JSON files to Grafana when:

1. The dashboard module starts up
2. `ceph dashboard update-grafana-dashboards` is run
3. A POST request is sent to `/api/grafana/update_dashboards`

b-ranto commented:
I have rebased, squashed the patches and made me and Zack the co-authors for the commit. I have also updated all the task names/titles to begin with lowercase.

b-ranto commented:
I have pushed a new version. The changes include a switch from `dict`s in `default/main.yml` to regular variables because it started to create problems after the rebase (and the rest of the code does it that way), use of both `mons` and `mgrs` for the MGR role in various places and another rebase.

b-ranto commented:
Latest changelog:
- we are now setting web.external-url for prometheus, this fixes an issue with bad links in the dashboard web UI for alerts
- ceph-grafana-dashboards is now being installed on the ceph-grafana node, not the ceph-mgr node
- this installs ceph-mgr-dashboard if dashboard is enabled, the dashboard was split into a separate package for Nautilus
- added ceph-common and ceph-config tasks to the ceph-grafana node (we need the repos there, too)

zmc commented:
@guits, @b-ranto Is there anything else that should be changed in this PR, or should we merge it? Thanks!

guits commented:
@zmc I guess we just need to allocate some times to do a final review.
I'll start to take a look this week.

epuertat commented:
Just for the case this gets merged sooner (https://github.com/ceph/ceph/pull/2691), it introduces a separate option for configuring SSL server port.

hoerup commented:
> Just for the case this gets merged sooner ([ceph/ceph#2691](https://github.com/ceph/ceph/pull/2691)), it introduces a separate option for configuring SSL server port.

I think you mean https://github.com/ceph/ceph/pull/26914 ??

b-ranto commented:
@guits Can we make any progress, here? Can you do at least a basic sanity check and merge if it looks ok? We (me and Zack) can handle the updates as necessary afterwards.

dsavineau commented:
@b-ranto could you amend your HEAD~1 commit and add a signoff to allow the CI to run ?
if you could rebase your branch on master too it will be great.

b-ranto commented:
@dsavineau Done and done.

b-ranto commented on infrastructure-playbooks/purge-docker-cluster.yml:
```diff
@@ -444,6 +449,132 @@
+      failed_when: false
+
+    - name: remove node-exporter container
+      docker_container:
```
The `purge-docker-cluster.yml` playbook does not seem to care about podman (all the other container removals there use `docker_container`), at least not yet. I am really not sure what to do about this.

b-ranto commented on roles/ceph-dashboard/tasks/configure_dashboard.yml:
```diff
@@ -0,0 +1,166 @@
+- name: disable SSL for dashboard
+  shell: |
+    {{ mgr_prefix }} ceph config set mgr mgr/dashboard/ssl false || \
+    {{ mgr_prefix }} ceph config-key set mgr/dashboard/ssl false
```
Yes, this is to support both luminous as well as mimic. The command has changed in between the releases and this seems to work reliably for both.

guits replied:
@b-ranto stable-4.0 won't support luminous and mimic

b-ranto commented on roles/ceph-dashboard/tasks/configure_dashboard.yml:
```diff
@@ -0,0 +1,166 @@
+  until: ac_result.rc == 0
+
+- name: set grafana url
+  command: "{{ mgr_prefix }} ceph dashboard set-grafana-api-url {{ dashboard_protocol }}://{{ groups['ceph-grafana'][0] }}:3000/"
```
We can't. We embed grafana in the dashboard and for that to work reasonably, you need to have the protocols synced between grafana and the dashboard.

b-ranto commented on roles/ceph-dashboard/tasks/configure_dashboard.yml:
```diff
@@ -0,0 +1,166 @@
+  changed_when: false
+
+- name: create radosgw system user
+  shell: "timeout 20 {{ mgr_prefix }} radosgw-admin user create --uid={{ dashboard_rgw_api_user_id }} --display-name='Ceph dashboard' --system"
```
The `radosgw-admin` is part of the `ceph-common` package. That package is always installed on the mgr as well as mon node.

b-ranto commented on roles/ceph-grafana/tasks/configure_grafana.yml:
```diff
@@ -0,0 +1,83 @@
+    # This is the UID used by the grafana container
+    owner: "{{ grafana_uid }}"
+    # This group is used by the grafana rpm
+    group: "grafana"
```
This is a hack to make it work with both grafana container and a grafana rpm. You can use gid to access the files from the regular machine and the uid to access it inside the container.

guits commented on infrastructure-playbooks/purge-cluster.yml:
```diff
@@ -42,12 +42,130 @@
     - "{{ nfs_group_name|default('nfss') }}"
     - "{{ client_group_name|default('clients') }}"
     - "{{ mgr_group_name|default('mgrs') }}"
+    - ceph-grafana
```
```suggestion
    - grafana
```

to keep consistency with other role names

b-ranto replied:
I suppose it should be `grafana-server` then.

guits commented on infrastructure-playbooks/purge-cluster.yml:
```diff
@@ -42,12 +42,130 @@
+    - "{{ client_group_name|default('clients') }}"
+    - "{{ mgr_group_name|default('mgrs') }}"
+    - agents
+    - ceph-grafana
```
```suggestion
    - grafana
```

guits commented on infrastructure-playbooks/purge-cluster.yml:
```diff
@@ -42,12 +42,130 @@
+
+
+- name: purge ceph-grafana
+  hosts: ceph-grafana
```
```suggestion
  hosts: grafana
```

guits commented on infrastructure-playbooks/purge-cluster.yml:
```diff
@@ -466,6 +585,7 @@
     - "{{ nfs_group_name|default('nfss') }}"
     - "{{ client_group_name|default('clients') }}"
     - "{{ mgr_group_name|default('mgrs') }}"
+    - ceph-grafana
```
```suggestion
    - grafana
```

guits commented on infrastructure-playbooks/purge-docker-cluster.yml:
```diff
@@ -444,6 +449,132 @@
+    - "{{ nfs_group_name|default('nfss') }}"
+    - "{{ mgr_group_name|default('mgrs') }}"
+    - agents
+    - ceph-grafana
```
```suggestion
    - grafana
```

guits commented on infrastructure-playbooks/purge-docker-cluster.yml:
```diff
@@ -444,6 +449,132 @@
+
+- name: purge ceph-grafana
+
+  hosts: ceph-grafana
```
```suggestion
  hosts: grafana
```

guits commented on roles/ceph-grafana/templates/datasources-ceph-dashboard.yml:
```diff
@@ -0,0 +1,26 @@
+  # <int> org id. will default to orgId 1 if not specified
+  orgId: 1
+  # <string> url
+  url: 'http://{{ groups["ceph-grafana"][0] }}:9090'
```
```suggestion
  url: 'http://{{ groups["grafana"][0] }}:9090'
```

guits commented on roles/ceph-prometheus/tasks/setup_container.yml:
```diff
@@ -0,0 +1,73 @@
+---
+- name: include ceph-container-common
+  include_role:
```
as discussed on IRC, I think it looks a bit agressive to have to run ceph-container-common just to get docker installed in a non containerized deployment, especially when we know the preferred way to run containers in 4.0 will be podman.

guits commented on site-container.yml.sample:
```diff
@@ -13,6 +13,7 @@
   - iscsigws
   - iscsi-gws # for backward compatibility only!
   - mgrs
+  - ceph-grafana
```
```suggestion
  - grafana
```

guits commented on site-container.yml.sample:
```diff
@@ -482,3 +510,45 @@
       run_once: true
       when: not ceph_status.failed
+
+- hosts: ceph-grafana
```
```suggestion
- hosts: grafana
```

guits commented on site.yml.sample:
```diff
@@ -13,6 +13,7 @@
   - mgrs
   - iscsigws
   - iscsi-gws # for backward compatibility only!
+  - ceph-grafana
```
```suggestion
  - grafana
```

guits commented on site.yml.sample:
```diff
@@ -485,3 +516,45 @@
       run_once: true
       when: not ceph_status.failed
+
+- hosts: ceph-grafana
```
```suggestion
- hosts: grafana
```

guits commented on roles/ceph-dashboard/tasks/configure_dashboard.yml:
```diff
@@ -0,0 +1,166 @@
+    dest: "/etc/ceph/ceph-dashboard.key"
+    owner: root
+    group: root
+    mode: 0644
```
@b-ranto could you clarify if `0644` is a requirement or we can go with `0400` ?

guits commented on infrastructure-playbooks/purge-cluster.yml:
```diff
@@ -42,12 +42,130 @@
+      failed_when: false
+
+    - name: remove node-exporter container
+      docker_container:
```
this isn't a topic specific to ceph-dashboard, let's address this in a dedicated PR.

b-ranto commented on roles/ceph-grafana/templates/grafana.repo:
```diff
@@ -0,0 +1,9 @@
+[grafana]
```
This can be removed since we are only deploying grafana in a container, nowadays.

b-ranto commented on roles/ceph-grafana/files/grafana.list:
```diff
@@ -0,0 +1 @@
+deb https://packagecloud.io/grafana/stable/debian/ jessie main
```
We don't need this file anymore, we should remove it. We are deploying containers-only grafana.

b-ranto commented on roles/ceph-grafana/tasks/main.yml:
```diff
@@ -0,0 +1,15 @@
+---
+- name: install ceph-grafana-dashboards package on RedHat or SUSE
+  package:
+    name: ceph-grafana-dashboards
```
Yeah, this is an old code, we should remove it.

b-ranto commented on site-container.yml.sample:
```diff
@@ -482,3 +510,45 @@
+        name: ceph-grafana
+      when: dashboard_enabled
+
+- hosts: '{{ (groups["mgrs"] | default(groups["mons"]))[0] }}'
```
We are doing this either on first mgr node or first mon node, we use `(mgrs | default(mons))[0]`, here.

b-ranto commented on roles/ceph-mgr/tasks/pre_requisite.yml:
```diff
@@ -7,6 +7,25 @@
   until: result is succeeded
   when: ansible_os_family in ['RedHat', 'Suse']
 
+- name: install ceph-grafana-dashboards package on RedHat or SUSE
```
Sadly, this package is not available in Ubuntu. The upstream packaging changes were only made for RedHat/SUSE.

b-ranto commented on roles/ceph-prometheus/templates/alertmanager.yml:
```diff
@@ -0,0 +1,15 @@
+- name: 'ceph-dashboard'
+  webhook_configs:
+{% for host in groups['mgrs'] | default(groups['mons']) %}
+  - url: 'http://{{ host }}:{{ dashboard_port }}/api/prometheus_receiver'
```
Yes, we should use the `dashboard_protocol`, here.

guits commented:
jenkins test dev-centos-non_container-dashboard

guits commented:
jenkins test dev-centos-container-dashboard

guits commented:
jenkins test dev-centos-non_container-dashboard

guits commented:
jenkins test dev-centos-container-dashboard

guits commented:
jenkins test dev-centos-non_container-dashboard

guits commented:
jenkins test dev-centos-container-dashboard

guits commented:
jenkins test dev-centos-container-dashboard

guits commented:
jenkins test dev-centos-non_container-dashboard
